src/test/java/states/InvalidOpCodeStateTest.java

```diff
@@ -0,0 +1,185 @@
+            DatagramPacket invalidPacket = new DatagramPacket(invalidOPData, invalidOPData.length, serverSocketAddress);
+            Mockito.when(socket.receive())
+                    .thenReturn(expectedPacket)
+                    .thenThrow(new RuntimeException("TEST EXCEPTION"));
+
+            state.setServerWorkerPort(3000);
+            state.setClientAddress(serverSocketAddress);
+            state.execute();
+
+            InvalidPacketMatcher matcher = new InvalidPacketMatcher(invalidPacket);
+            Mockito.verify(socket, Mockito.times(0)).forwardPacket(Mockito.eq(expectedPacket), Mockito.eq(serverAddress), Mockito.eq(3000));
+            Mockito.verify(socket, Mockito.times(1)).forwardPacket(Mockito.argThat(matcher), Mockito.eq(serverAddress), Mockito.eq(3000));
+        } catch (IOException e) {
+            Assert.fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testInvalidACK() {
+        try {
+            checker = new ErrorChecker(Message.MessageType.ACK, 1);
+            state.setErrorChecker(checker);
+
+            byte[] expectedACKBytes = new AckMessage(1).toByteArray();
+            invalidOPData = duplicateDataInvOp(expectedACKBytes);
+            
+            DatagramPacket expectedPacket = new DatagramPacket(expectedACKBytes, expectedACKBytes.length, serverSocketAddress);
+            DatagramPacket invalidPacket = new DatagramPacket(invalidOPData, invalidOPData.length, serverSocketAddress);
+            Mockito.when(socket.receive())
+                    .thenReturn(expectedPacket)
+                    .thenThrow(new RuntimeException("TEST EXCEPTION"));
+
+            state.setServerWorkerPort(3000);
+            state.setClientAddress(serverSocketAddress);
+            state.execute();
+
+            InvalidPacketMatcher matcher = new InvalidPacketMatcher(invalidPacket);
+            Mockito.verify(socket, Mockito.times(0)).forwardPacket(Mockito.eq(expectedPacket), Mockito.eq(serverAddress), Mockito.eq(3000));
+            Mockito.verify(socket, Mockito.times(1)).forwardPacket(Mockito.argThat(matcher), Mockito.eq(serverAddress), Mockito.eq(3000));
+        } catch (IOException e) {
+            Assert.fail(e.getMessage());
+        }
+    }
+    
+    public byte[] duplicateDataInvOp(byte[] arr) {
+    	byte [] arrVal = new byte[arr.length];
+    	for(int i = 0; i < arr.length; i++) {
+    		if( i != 1) {
+    			arrVal[i] = arr[i];
+    		} else {
+    			arrVal[i] = 7;
+    		}
+    	}
+    	return arrVal;
+    }
+
+}
+
+class InvalidPacketMatcher implements ArgumentMatcher<DatagramPacket> {
+	private DatagramPacket packet;
+	
+	public InvalidPacketMatcher(DatagramPacket packet) {
+		this.packet = packet;
+	}
+	
+	@Override
+	public boolean matches(DatagramPacket p) {
+		byte [] packetData = packet.getData();
+		byte [] pData = p.getData();
+		if( p.getLength() == packet.getLength() ) {
```

Comment:
Use this utility function? 

return Arrays.equals(p.getData(),this.packet.getData());